[Merged by Bors] - feat(number_theory/cyclotomic/zeta): add lemmas

riccardobrasca opened the pull request:
Lemmas about the norm of `ζ - 1`.

From flt-regular.

- [x] depends on: #11941

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

riccardobrasca committed:
fix doc

riccardobrasca committed:
better implicit variables

riccardobrasca committed:
add lemmas

riccardobrasca committed:
add doc

ericrbg reviewed (approved):
the fact that these are all separate lemmas are evidence against maths being beautiful :b

lgtm, I wish we had a `ne_zero` pow but that also has to deal with coes and stuff and that doesn't work cleanly... I dislike pnat and the fact we have to use it :(

what would be nice is a tactic that changes statements about subtypes about statements about the big type, but oh well

riccardobrasca committed:
add prime_pow_ne_two.norm_zeta_sub_one

ericrbg commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+
+/-- If `is_prime_pow (n : ℕ)`, `n ≠ 2` and `irreducible (cyclotomic n K)` (in particular for
+`K = ℚ`), then the norm of `zeta n K L - 1` is `(n : ℕ).min_fac`. -/
+lemma prime_pow_ne_two.norm_zeta_sub_one {n : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
```
```suggestion
lemma is_prime_pow.norm_zeta_sub_one {n : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
```

for dot notation

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Eric Rodriguez <37984851+ericrbg@users.noreply.github.com>

riccardobrasca committed:
typo

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

ericrbg committed:
chore(number_theory/cyclotomic/zeta): generalize to primitive roots

ericrbg committed:
fix copyright line

I was oriignally splitting this, and then decided to not - the copyright line got caught in the crossfire, sorry!

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -5,11 +5,13 @@ Authors: Alex Best, Riccardo Brasca, Eric Rodriguez
 import ring_theory.adjoin.power_basis
 import ring_theory.norm
+import algebra.is_prime_pow
 
```
My hobby: sorting import statements.
```suggestion
import algebra.is_prime_pow
import number_theory.cyclotomic.basic
import ring_theory.adjoin.power_basis
import ring_theory.polynomial.cyclotomic.eval
import ring_theory.norm

```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+
+/-- If `is_prime_pow (n : ℕ)`, `n ≠ 2` and `irreducible (cyclotomic n K)` (in particular for
+`K = ℚ`), then the norm of `zeta n K L - 1` is `(n : ℕ).min_fac`. -/
+lemma is_prime_pow.norm_zeta_sub_one {n : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
```
Wouldn't `is_prime_pow.one_lt hn` imply that `n` is positive anyway, so we don't need so much coercing? Or is it easier to keep the whole API `pnat`-based?
```suggestion
lemma is_prime_pow.norm_zeta_sub_one {n : ℕ} {K : Type u} (L : Type v) [field K] [field L]
```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+end
+
+/-- If `irreducible (cyclotomic (p ^ k) K)` (in particular for `K = ℚ`) and `p` is an odd prime,
+then the norm of `zeta (p  ^ k) K L - 1` is `p`. -/
```
```suggestion
then the norm of `zeta (p ^ k) K L - 1` is `p`. -/
```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+
+/-- If `irreducible (cyclotomic (p ^ k) K)` (in particular for `K = ℚ`) and `p` is an odd prime,
+then the norm of `zeta (p  ^ k) K L - 1` is `p`. -/
+lemma prime_ne_two_pow.norm_zeta_sub_one {p : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
```
Similarly, here `p` is already positive because it's an odd prime.
```suggestion
lemma prime_ne_two_pow.norm_zeta_sub_one {p : ℕ} {K : Type u} (L : Type v) [field K] [field L]
```

ericrbg replied:
the issue here is that `is_cyclotomic_extension` takes a `set ℕ+`; I'm not sure if the coercion works with typeclass search and similar

Vierkantor replied:
I see. Then it's probably fine to keep `pnat` everywhere.

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+
+/-- If `irreducible (cyclotomic p K)` (in particular for `K = ℚ`) and `p` is an odd prime,
+then the norm of `zeta p K L - 1` is `p`. -/
+lemma prime_ne_two.norm_zeta_sub_one {p : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
```
```suggestion
lemma prime_ne_two.norm_zeta_sub_one {p : ℕ} {K : Type u} (L : Type v) [field K] [field L]
```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+lemma is_prime_pow.norm_zeta_sub_one {n : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
+  [ne_zero ((n : ℕ) : K)] (hn : is_prime_pow (n : ℕ)) [algebra K L]
+  [is_cyclotomic_extension {n} K L]
+  (hirr : irreducible (cyclotomic ((n : ℕ)) K)) (h : n ≠ 2) :
```
```suggestion
  (hirr : irreducible (cyclotomic (n : ℕ) K)) (h : n ≠ 2) :
```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+lemma prime_ne_two_pow.norm_zeta_sub_one {p : ℕ+} {K : Type u} (L : Type v) [field K] [field L]
+  [ne_zero ((p : ℕ) : K)] (k : ℕ) [hpri : fact (p : ℕ).prime] [algebra K L]
+  [is_cyclotomic_extension {p ^ (k + 1)} K L]
+  (hirr : irreducible (cyclotomic ((↑(p ^ (k + 1)) : ℕ)) K)) (h : p ≠ 2) :
```
```suggestion
  (hirr : irreducible (cyclotomic (↑(p ^ (k + 1)) : ℕ)) K) (h : p ≠ 2) :
```

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+    have := lt_of_le_of_ne hpri.1.two_le (by contrapose! h; exact coe_injective h.symm),
+    refine lt_of_lt_of_le this _,
+    nth_rewrite 0 [← pow_one ↑p],
+    refine pow_le_pow (one_le_of_lt this) le_add_self },
```
This part would probably be clearer as a `calc` block.

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

Vierkantor commented on src/number_theory/cyclotomic/zeta.lean:
```diff
@@ -176,6 +179,84 @@ begin
+  have : 2 < (2 ^ k : ℕ+),
+  { simp only [← coe_lt_coe, pnat.coe_bit0, one_coe, pow_coe],
+    nth_rewrite 0 [← pow_one 2],
+    refine pow_lt_pow one_lt_two (lt_of_lt_of_le one_lt_two hk) },
```
Shouldn't this just be:
```suggestion
    exact pow_lt_pow one_lt_two (lt_of_lt_of_le one_lt_two hk) },
```

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/zeta.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
golf

riccardobrasca committed:
add norm_primitive_root_eq_one

riccardobrasca committed:
Merge remote-tracking branch 'origin/zeta_refactor_again' into RB/norm_zeta_sub_one

ericrbg committed:
many doc fixups

ericrbg committed:
remove by begin

riccardobrasca committed:
add lemmas

riccardobrasca committed:
doc

riccardobrasca committed:
explicit variable

riccardobrasca committed:
Merge branch 'zeta_refactor_again' into RB/norm_zeta_sub_one

ericrbg committed:
oopsies

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

riccardobrasca committed:
Merge branch 'zeta_refactor_again' into RB/norm_zeta_sub_one

ericrbg committed:
much better message!

Co-authored-by: Riccardo Brasca <riccardo.brasca@gmail.com>

riccardobrasca committed:
Merge branch 'zeta_refactor_again' into RB/norm_zeta_sub_one

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

riccardobrasca committed:
useless

riccardobrasca committed:
ops

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,155 @@ begin
-end norm
+/-- If `is_prime_pow (n : ℕ)`, `n ≠ 2` and `irreducible (cyclotomic n K)` (in particular for
+`K = ℚ`), then the norm of `ζ - 1` is `(n : ℕ).min_fac`. -/
+lemma is_prime_pow_sub_one_norm [ne_zero ((n : ℕ) : K)] (hn : is_prime_pow (n : ℕ))
```
```suggestion
lemma is_prime_pow.sub_one_norm [ne_zero ((n : ℕ) : K)] (hn : is_prime_pow (n : ℕ))
```

turns on dot notation

ericrbg replied:
Oh, this was from an old review but for some reason it didn't delete. It's wrong!

riccardobrasca committed:
lint

riccardobrasca committed:
Merge branch 'master' into RB/norm_zeta_sub_one

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
-end norm
+/-- If `is_prime_pow (n : ℕ)`, `n ≠ 2` and `irreducible (cyclotomic n K)` (in particular for
+`K = ℚ`), then the norm of `ζ - 1` is `(n : ℕ).min_fac`. -/
+lemma is_prime_pow_sub_one_norm (hn : is_prime_pow (n : ℕ)) [is_cyclotomic_extension {n} K L]
```
I'm gonna suggest a different naming for this because I really don't like these:

```suggestion
lemma sub_one_norm.is_prime_pow (hn : is_prime_pow (n : ℕ)) [is_cyclotomic_extension {n} K L]
```

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+
+/-- If `irreducible (cyclotomic (p ^ (k + 1)) K)` (in particular for `K = ℚ`) and `p` is an odd
+prime, then the norm of `ζ - 1` is `p`. -/
+lemma prime_ne_two_pow.sub_one_norm {p : ℕ+} [ne_zero ((p : ℕ) : K)] (k : ℕ)
```
```suggestion
lemma sub_one_norm.pow_prime {p : ℕ+} [ne_zero ((p : ℕ) : K)] (k : ℕ)
```

I don't think you need to say that you don't accept 2 here, if anything in the docstring and link to the other decl

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+
+/-- If `irreducible (cyclotomic p K)` (in particular for `K = ℚ`) and `p` is an odd prime,
+then the norm of `ζ - 1` is `p`. -/
+lemma prime_ne_two.sub_one_norm {p : ℕ+} [ne_zero ((p : ℕ) : K)] [hpri : fact (p : ℕ).prime]
```
```suggestion
lemma sub_one_norm.prime {p : ℕ+} [ne_zero ((p : ℕ) : K)] [hpri : fact (p : ℕ).prime]
```

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+
+/-- If `irreducible (cyclotomic (2 ^ k) K)` (in particular for `K = ℚ`) and `k` is at least `2`,
+then the norm of `ζ - 1` is `2`. -/
+lemma two_pow.sub_one_norm [ne_zero (2 : K)] {k : ℕ} (hk : 2 ≤ k)
```
```suggestion
lemma sub_one_norm.pow_two [ne_zero (2 : K)] {k : ℕ} (hk : 2 ≤ k)
```

ericrbg commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+
+/-- If `K` is linearly ordered (in particular for `K = ℚ`), the norm of `zeta n K L` is `1`
+if `n` is odd. -/
+lemma norm_zeta_eq_one (K : Type u) [linear_ordered_field K] (L : Type v) [field L] [algebra K L]
```
I don't really see why we need all the lemmas specifically for `zeta`, `(zeta_primitive_root ...).` should be fairly natural in the proof

riccardobrasca replied:
I would say it's reasonable to have them, since we have an explicit definition for `zeta`, but I don't have any strong opinion.

Vierkantor replied:
The `haveI` in the proof below suggests to me that there will be enough of a hassle to not make inlining worth it. So I vote for keeping them.

ericrbg reviewed (approved):
apart from naming quibbles lgtm

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Eric Rodriguez <37984851+ericrbg@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Eric Rodriguez <37984851+ericrbg@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Eric Rodriguez <37984851+ericrbg@users.noreply.github.com>

riccardobrasca committed:
fix build

Vierkantor commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+/-- If `irreducible (cyclotomic (2 ^ k) K)` (in particular for `K = ℚ`) and `k` is at least `2`,
+then the norm of `ζ - 1` is `2`. -/
+lemma sub_one_norm.pow_two [ne_zero (2 : K)] {k : ℕ} (hk : 2 ≤ k)
+  [is_cyclotomic_extension {2 ^ k} K L] (hζ: is_primitive_root ζ (2 ^ k))
```
```suggestion
  [is_cyclotomic_extension {2 ^ k} K L] (hζ : is_primitive_root ζ (2 ^ k))
```

Vierkantor commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+/-- If `irreducible (cyclotomic (p ^ (k + 1)) K)` (in particular for `K = ℚ`) and `p` is an odd
+prime, then the norm of `ζ - 1` is `p`. -/
+lemma prime_ne_two_pow.sub_one_norm {p : ℕ+} [ne_zero ((p : ℕ) : K)] (k : ℕ)
+  (hζ: is_primitive_root ζ ↑(p ^ (k + 1))) [hpri : fact (p : ℕ).prime]
```
```suggestion
  (hζ : is_primitive_root ζ ↑(p ^ (k + 1))) [hpri : fact (p : ℕ).prime]
```

Vierkantor commented on src/number_theory/cyclotomic/primitive_roots.lean:
```diff
@@ -193,6 +202,154 @@ begin
+  obtain ⟨k, hk⟩ : ∃ k, ((n : ℕ).factorization) (n : ℕ).min_fac = k + 1 := exists_eq_succ_of_ne_zero
+      (((n : ℕ).factorization.mem_support_to_fun (n : ℕ).min_fac).1
+      $ factor_iff_mem_factorization.2 $ (mem_factors (is_prime_pow.ne_zero hn)).2 ⟨hprime.out,
+      min_fac_dvd _⟩),
```
Indenting like this makes the structure a bit clearer:
```suggestion
  obtain ⟨k, hk⟩ : ∃ k, ((n : ℕ).factorization) (n : ℕ).min_fac = k + 1 :=
    exists_eq_succ_of_ne_zero (((n : ℕ).factorization.mem_support_to_fun (n : ℕ).min_fac).1 $
      factor_iff_mem_factorization.2 $ (mem_factors (is_prime_pow.ne_zero hn)).2
        ⟨hprime.out, min_fac_dvd _⟩),
```

Vierkantor reviewed (approved):
Looks good to me with some minor quibbles!

bors d+

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

riccardobrasca committed:
Update src/number_theory/cyclotomic/primitive_roots.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>
